### PR TITLE
Add EventSource event method generation

### PR DESCRIPTION
**Pull request by Cricle**
Contributes to https://github.com/dotnet/runtime/issues/107069

- [x] Final result

```csharp
//Writing code...

[EventSourceEventGenerate]
internal sealed unsafe /*It's mandatory or in method write unsafe*/ partial class RuntimeEventSource : EventSource
{
    [Event(1)]
    public partial void GoHome(string address,double usedTime);
}


//Will generated
public void GoHome(string address,double usedTime)
{
    address ??= System.String.Empty;
    global::System.Diagnostics.Tracing.EventSource.EventData* datas = stackalloc global::System.Diagnostics.Tracing.EventSource.EventData[2];
    datas[0] = new global::System.Diagnostics.Tracing.EventSource.EventData
    {
        //Use unsafe method because it is faster than fixed, and easily to generate
        DataPointer = (nint)global::System.Runtime.CompilerServices.Unsafe.AsPointer(ref global::System.Runtime.InteropServices.MemoryMarshal.GetReference(global::System.MemoryExtensions.AsSpan(address)),
        Size = (address.Length + 1) * sizeof(char)
    };
    datas[1] = new global::System.Diagnostics.Tracing.EventSource.EventData
    {
        DataPointer = (nint)(&usedTime),
        Size = sizeof(double)
    };
    WriteEventWithRelatedActivityIdCore(1, null, 2, datas);
    OnRaise(i, address);
}

[global::System.Diagnostics.Tracing.NonEventAttribute]
partial void OnRaise(int i, string address);
```

### CODE REVIEW

**Comment by Cricle** [comment]
But I don't how to start the generation test. 

Please let me know if you need additional tests, or if you need to modify the way you test now

File: src/libraries/System.Diagnostics.Tracing/ref/System.Diagnostics.Tracing.cs [hunk 1]
```diff
@@ -14,6 +14,11 @@ public enum EventActivityOptions
         Recursive = 4,
         Detachable = 8,
     }
+    [AttributeUsage(System.AttributeTargets.Class)]
+    public sealed class EventSourceEventGenerateAttribute : Attribute
```
Comment (am11) [inline src/libraries/System.Diagnostics.Tracing/ref/System.Diagnostics.Tracing.cs:18]: Should we call it `GeneratedEventSourceEventAttribute`, akin to `GeneratedRegex`?

Comment (am11) [reply src/libraries/System.Diagnostics.Tracing/ref/System.Diagnostics.Tracing.cs:18]: Or `GeneratedTracingEventAttribute`

**Comment by tommcdon** [comment]
@noahfalk @tarekgh would you mind reviewing this PR?

**Comment by tarekgh** [comment]
@tommcdon @Cricle this need to go through design review before accepting it. https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md

**Comment by Cricle** [comment]
> @tommcdon @Cricle this need to go through design review before accepting it. https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md

Would I need to create an 'API Proposal' issue link this PR?

**Comment by tarekgh** [comment]
@Cricle you may update https://github.com/dotnet/runtime/issues/107069 and use it for the design review. I would suggest closing this PR till the design review is done because this will depend on when the design review will be scheduled. Thanks for your help here!

**Comment by Cricle** [comment]
@tarekgh @tommcdon I has been update this issure #107069

**Comment by tarekgh** [comment]
I'll let @noahfalk look at the proposal and mark the issue as ready for review when it is ready. @Cricle do you mind closing this PR till the request get reviewed and approved?

**Comment by noahfalk** [comment]
Thanks @Cricle - I wasn't able to get to this today but its on my todo list to look at tomorrow.

**Pull request closed without merging**